File: knitro-10.3.0-z-Linux-64/examples/C/problemQCQP.c
/*******************************************************/
/* Copyright (c) 2016 by Artelys                       */
/* All Rights Reserved                                 */
/*******************************************************/

/*++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 *  This file contains routines to implement problemDef.h for a small
 *  QCQP (quadratically constrained quadratic programming) test problem.
 *
 *  min   1000 - x0^2 - 2 x1^2 - x2^2 - x0 x1 - x0 x2
 *  s.t.  8 x0 + 14 x1 + 7 x2 - 56 = 0
 *        x0^2 + x1^2 + x2^2 - 25 >= 0
 *        x0 >= 0, x1 >= 0, x2 >= 0
 *
 *  The start point (2, 2, 2) converges to the minimum at (0, 0, 8),
 *  with final objective = 936.0.  From a different start point,
 *  Knitro may converge to an alternate local solution at (7, 0, 0),
 *  with objective = 951.0.
 *++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*/


#include "problemDef.h"
#include "knitro.h"


/*------------------------------------------------------------------*/
/*     FUNCTION getProblemSizes                                     */
/*------------------------------------------------------------------*/
/** Define sizes for the problem.
 */
void  getProblemSizes (int *  const  n,
                       int *  const  m,
                       int *  const  nnzJ,
                       int *  const  nnzH)
{
    *n = 3;
    *m = 2;
    *nnzJ = 6;
    *nnzH = 5;

    return;
}


/*------------------------------------------------------------------*/
/*     FUNCTION loadProblemArrays                                   */
/*------------------------------------------------------------------*/
/** Load Knitro arrays for the problem.
 *  The function assumes array memory has been allocated to match
 *  the problem sizes.
 *
 *  For more information about the arguments, refer to the Knitro
 *  manual, especially the section on the Callable Library.
 */
void  getProblemData (int    * const  objType,
                      int    * const  objGoal,
                      double * const  xLoBnds,
                      double * const  xUpBnds,
                      double * const  xInitial,
                      int    * const  cType,
                      double * const  cLoBnds,
                      double * const  cUpBnds,
                      int    * const  jacIndexVars,
                      int    * const  jacIndexCons,
                      int    * const  hessRows,
                      int    * const  hessCols,
                      int    * const  objFnType,
                      int    * const  xType,    
                      int    * const  cFnType)
{
    /*---- DEFINE THE OBJECTIVE FUNCTION AND VARIABLE BOUNDS. */
    *objType = KTR_OBJTYPE_QUADRATIC;
    *objGoal = KTR_OBJGOAL_MINIMIZE;
    xLoBnds[0] = 0.0;
    xUpBnds[0] = KTR_INFBOUND;
    xLoBnds[1] = 0.0;
    xUpBnds[1] = KTR_INFBOUND;
    xLoBnds[2] = 0.0;
    xUpBnds[2] = KTR_INFBOUND;
    xInitial[0] = 2.0;
    xInitial[1] = 2.0;
    xInitial[2] = 2.0;

    /*---- DEFINE THE CONSTRAINT FUNCTIONS. */
    cType[0] = KTR_CONTYPE_LINEAR;
    cType[1] = KTR_CONTYPE_QUADRATIC;
    cLoBnds[0] = 0.0;
    cUpBnds[0] = 0.0;
    cLoBnds[1] = 0.0;
    cUpBnds[1] = KTR_INFBOUND;

    /*---- PROVIDE FIRST DERIVATIVE STRUCTURAL INFORMATION. */
    jacIndexCons[0] = 0;
    jacIndexVars[0] = 0;
    jacIndexCons[1] = 0;
    jacIndexVars[1] = 1;
    jacIndexCons[2] = 0;
    jacIndexVars[2] = 2;
    jacIndexCons[3] = 1;
    jacIndexVars[3] = 0;
    jacIndexCons[4] = 1;
    jacIndexVars[4] = 1;
    jacIndexCons[5] = 1;
    jacIndexVars[5] = 2;

    /*---- PROVIDE SECOND DERIVATIVE STRUCTURAL INFORMATION;
     *---- ONLY THE NONZEROES OF THE UPPER TRIANGLE OF THE HESSIAN. */
    hessRows[0] = 0;
    hessCols[0] = 0;
    hessRows[1] = 0;
    hessCols[1] = 1;
    hessRows[2] = 0;
    hessCols[2] = 2;
    hessRows[3] = 1;
    hessCols[3] = 1;
    hessRows[4] = 2;
    hessCols[4] = 2;

    return;
}


/*------------------------------------------------------------------*/
/*     FUNCTION computeFC                                           */
/*------------------------------------------------------------------*/
/** Compute the function and constraint values at x.
 *
 *  For more information about the arguments, refer to the Knitro
 *  manual, especially the section on the Callable Library.
 */
double  computeFC (const double * const  x,
                         double * const  c)
{
    double  dObj;

    dObj = 1.0e3 - x[0]*x[0] - 2.0e0*x[1]*x[1] - x[2]*x[2]
                 - x[0]*x[1] - x[0]*x[2];

    /*---- LINEAR EQUALITY CONSTRAINT. */
    c[0] = 8.0e0*x[0] + 14.0e0*x[1] + 7.0e0*x[2] - 56.0e0;

    /*---- QUADRATIC INEQUALITY CONSTRAINTS. */
    c[1] = x[0]*x[0] + x[1]*x[1] + x[2]*x[2] - 25.0e0;

    return( dObj );
}


/*------------------------------------------------------------------*/
/*     FUNCTION computeGA                                           */
/*------------------------------------------------------------------*/
/** Compute the function and constraint first derivatives at x.
 *
 *  For more information about the arguments, refer to the Knitro
 *  manual, especially the section on the Callable Library.
 */
void  computeGA (const double * const  x,
                       double * const  objGrad,
                       double * const  jac)
{
    /*---- GRADIENT OF THE OBJECTIVE FUNCTION. */
    objGrad[0] = -2.0e0*x[0] - x[1] - x[2];
    objGrad[1] = -4.0e0*x[1] - x[0];
    objGrad[2] = -2.0e0*x[2] - x[0];

    /*---- GRADIENT OF THE FIRST CONSTRAINT, c[0]. */
    jac[0] =  8.0e0;
    jac[1] = 14.0e0;
    jac[2] =  7.0e0;


    /*---- GRADIENT OF THE SECOND CONSTRAINT, c[1]. */
    jac[3] = 2.0e0*x[0];
    jac[4] = 2.0e0*x[1];
    jac[5] = 2.0e0*x[2];

    return;
}


/*------------------------------------------------------------------*/
/*     FUNCTION computeH                                            */
/*------------------------------------------------------------------*/
/** Compute the Hessian of the Lagrangian at x and lambda.
 *
 *  For more information about the arguments, refer to the Knitro
 *  manual, especially the section on the Callable Library.
 */
void  computeH (const double * const  x,
                const double          objScaler,
                const double * const  lambda,
                      double * const  hess)
{
    hess[0] = -2.0e0*objScaler + 2.0e0*lambda[1];
    hess[1] = -1.0e0*objScaler;
    hess[2] = -1.0e0*objScaler;
    hess[3] = -4.0e0*objScaler + 2.0e0*lambda[1];
    hess[4] = -2.0e0*objScaler + 2.0e0*lambda[1];

    return;
}


/*------------------------------------------------------------------*/
/*     FUNCTION computeHV                                           */
/*------------------------------------------------------------------*/
/** Compute the Hessian of the Lagrangian times vector at x and lambda.
 *  Return the result in vector.
 *
 *  For more information about the arguments, refer to the Knitro
 *  manual, especially the section on the Callable Library.
 */
void  computeHV (const double * const  x,
                 const double          objScaler,
                 const double * const  lambda,
                       double * const  vector)
{
    double  tmp[3];
    int     i;

    tmp[0] = (-2.0e0*objScaler + 2.0e0*lambda[1])*vector[0]
             - objScaler*vector[1] - objScaler*vector[2];
    tmp[1] = -objScaler*vector[0] + (-4.0e0*objScaler + 2.0e0*lambda[1])*vector[1];
    tmp[2] = -objScaler*vector[0] + (-2.0e0*objScaler + 2.0e0*lambda[1])*vector[2];

    for (i = 0; i < 3; i++)
        vector[i] = tmp[i];

    return;
}


/*----- End of source code -----------------------------------------*/

File: knitro-10.3.0-z-Linux-64/examples/Fortran/problemQCQP.f
C *******************************************************
C * Copyright (c) 2015 by Artelys                       *
C * All Rights Reserved                                 *
C *******************************************************

C+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
C  This file contains routines to define a small QCQP (quadratically
C  constrained quadratic programming) test problem.
C
C  min   1000 - x0^2 - 2 x1^2 - x2^2 - x0 x1 - x0 x2
C  s.t.  8 x0 + 14 x1 + 7 x2 - 56 = 0
C        x0^2 + x1^2 + x2^2 - 25 >= 0
C        x0 >= 0, x1 >= 0, x2 >= 0
C
C  The start point (2, 2, 2) converges to the minimum at (0, 0, 8),
C  with final objective = 936.0.  From a different start point,
C  Knitro may converge to an alternate local solution at (7, 0, 0),
C  with objective = 951.0.
C+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


C---------------------------------------------------------------------
C     FUNCTION get_problem_sizes
C---------------------------------------------------------------------

      subroutine get_problem_sizes (n, m, nnzj, nnzh)

      implicit none

      integer  n, m, nnzj, nnzh


      n = 3
      m = 2
      nnzj = 6
      nnzh = 5

      return
      end


C---------------------------------------------------------------------
C     FUNCTION load_problem_def
C---------------------------------------------------------------------

      subroutine load_problem_def (n, m, nnzj, nnzh,
     $                             objtype, objgoal,
     $                             xlobnds, xupbnds, xinitial,
     $                             ctype, clobnds, cupbnds,
     $                             jacindvar, jacindcon,
     $                             hrow, hcol)

      implicit none

      integer           n, m, nnzj, nnzh
      integer           objtype, objgoal
      double precision  xlobnds(n), xupbnds(n), xinitial(n)
      integer           ctype(m)
      double precision  clobnds(m), cupbnds(m)
      integer           jacindvar(nnzj), jacindcon(nnzj)
      integer           hrow(nnzh), hcol(nnzh)


C---- LOCAL VARIABLES.

      integer           i

C---- THE PARAMETER infbound MUST BE GREATER THAN OR EQUAL TO THE KNITRO
C---- PARAMETER KTR_INFBOUND DEFINED IN THE HEADER FILE "knitro.h".
C---- KNITRO TREATS ALL VALUES GREATER THAN OR EQUAL IN MAGNITUDE TO
C---- KTR_INFBOUND AS INFINITE.
      double precision  infbound
      parameter        (infbound = 1.d+20)


C---- THESE CONSTANTS ARE DEFINED IN "knitro.h":
C----   KTR_OBJGOAL_MINIMIZE = 0
C----   KTR_OBJGOAL_MAXIMIZE = 1
C----   KTR_OBJTYPE_GENERAL   = 0
C----   KTR_OBJTYPE_LINEAR    = 1
C----   KTR_OBJTYPE_QUADRATIC = 2
C----   KTR_CONTYPE_GENERAL   = 0
C----   KTR_CONTYPE_LINEAR    = 1
C----   KTR_CONTYPE_QUADRATIC = 2
      objgoal = 0
      objtype = 2
      ctype(1) = 1
      ctype(2) = 2

C---- SET VARIABLE BOUNDS AND INITIAL VALUES.
      do i=1,n
          xinitial(i) = 2.0
          xlobnds(i) = 0.0
          xupbnds(i) = infbound
      end do

C---- SET CONSTRAINT BOUNDS ( clobnd <= c(x) <= cupbnd ).
      clobnds(1) = 0.0
      cupbnds(1) = 0.0
      clobnds(2) = 0.0
      cupbnds(2) = infbound

C---- PROVIDE FIRST DERIVATIVE STRUCTURAL INFORMATION.
C---- NUMBER INDICES FROM ZERO ACCORDING TO THE C LANGUAGE CONVENTION,
C---- BECAUSE THE KNITRO SOLVER IS WRITTEN IN C.
      jacindcon(1) = 0
      jacindvar(1) = 0
      jacindcon(2) = 0
      jacindvar(2) = 1
      jacindcon(3) = 0
      jacindvar(3) = 2
      jacindcon(4) = 1
      jacindvar(4) = 0
      jacindcon(5) = 1
      jacindvar(5) = 1
      jacindcon(6) = 1
      jacindvar(6) = 2

C---- PROVIDE SECOND DERIVATIVE STRUCTURAL INFORMATION;
C---- ONLY THE NONZEROES OF THE UPPER TRIANGLE OF THE HESSIAN.
C---- NUMBER INDICES FROM ZERO ACCORDING TO THE C LANGUAGE CONVENTION,
C---- BECAUSE THE KNITRO SOLVER IS WRITTEN IN C.
      hrow(1) = 0
      hcol(1) = 0
      hrow(2) = 0
      hcol(2) = 1
      hrow(3) = 0
      hcol(3) = 2
      hrow(4) = 1
      hcol(4) = 1
      hrow(5) = 2
      hcol(5) = 2

      return
      end


C---------------------------------------------------------------------
C     FUNCTION compute_fc
C---------------------------------------------------------------------
C  COMPUTE THE FUNCTION AND CONSTRAINT VALUES AT x.

      subroutine compute_fc (x, obj, c)

      implicit none

      double precision  x(*), obj, c(*)


      obj = 1.0d+03 - x(1)**2 - 2.0d+00*x(2)**2
     $      - x(3)**2 - x(1)*x(2) - x(1)*x(3)


C---- LINEAR EQUALITY CONSTRAINT.
      c(1) = 8.0d+00*x(1) + 14.0d+00*x(2) + 7.0d+00*x(3) - 56.0d+00

C---- QUADRATIC INEQUALITY CONSTRAINT.
      c(2) = x(1)**2 + x(2)**2 + x(3)**2 - 25.0d+00

      return 
      end


C---------------------------------------------------------------------
C     FUNCTION compute_ga
C---------------------------------------------------------------------
C  COMPUTE THE FUNCTION AND CONSTRAINT FIRST DERIVATIVES AT x.

      subroutine compute_ga (x, objgrad, jac)

      implicit none

      double precision  x(*), objgrad(*), jac(*)


C---- GRADIENT OF THE OBJECTIVE FUNCTION.
      objgrad(1) = -2.0d0*x(1) - x(2) - x(3)
      objgrad(2) = -4.0d0*x(2) - x(1)
      objgrad(3) = -2.0d0*x(3) - x(1)

C---- GRADIENT OF THE FIRST CONSTRAINT, c(1).
      jac(1) =  8.0d0
      jac(2) = 14.0d0
      jac(3) =  7.0d0

C---- GRADIENT OF THE SECOND CONSTRAINT, c(2).
      jac(4) = 2.0d0*x(1)
      jac(5) = 2.0d0*x(2)
      jac(6) = 2.0d0*x(3)

      return
      end


C---------------------------------------------------------------------
C     FUNCTION compute_h
C---------------------------------------------------------------------
C  COMPUTE THE HESSIAN OF THE LAGRANGIAN AT x AND lambda.

      subroutine compute_h (x, objscaler, lambda, hess)

      implicit none

      double precision  objscaler
      double precision x(*), lambda(*), hess(*)

      hess(1) = -2.0d0*objscaler + 2.0d0*lambda(2)
      hess(2) = -1.0d0*objscaler
      hess(3) = -1.0d0*objscaler
      hess(4) = -4.0d0*objscaler + 2.0d0*lambda(2)
      hess(5) = -2.0d0*objscaler + 2.0d0*lambda(2) 

      return
      end


C---------------------------------------------------------------------
C     FUNCTION compute_hv
C---------------------------------------------------------------------
C  COMPUTE THE HESSIAN OF THE LAGRANGIAN TIMES vector AT x AND lambda.
C  RETURN THE RESULT IN vector.

      subroutine compute_hv (x, objscaler, lambda, vector)

      implicit none

      double precision  objscaler
      double precision  x(*), lambda(*), vector(*)

      integer           i
      double precision  tmp(3)


      tmp(1) = (-2.0d0*objscaler + 2.0d0*lambda(2))*vector(1)
     $         - objscaler*vector(2) - objscaler*vector(3)
      tmp(2) = -objscaler*vector(1) + 
     $         (-4.0d0*objscaler + 2.0d0*lambda(2))*vector(2)
      tmp(3) = -objscaler*vector(1) + 
     $         (-2.0d0*objscaler + 2.0d0*lambda(2))*vector(3)

      do i=1,3
          vector(i) = tmp(i)
      end do      

      return
      end


C----- End of source code -------------------------------------------
